src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressionCommentFilterTest.java

```diff
@@ -649,4 +619,71 @@ private static List<Comparable<Object>> getTagsAfterExecution(SuppressionComment
         return TestUtil.getInternalState(filter, "tags");
     }
 
+
+    @Test
+    public void testSuppressByCheck() throws Exception {
+        final String[] suppressedViolation = {
+            "27:9: "
+                + getCheckMessage(AbstractNameCheck.class,
+                    MSG_INVALID_PATTERN, "line_length", "^[a-z][a-zA-Z0-9]*$"),
+        };
+        final String[] expectedViolation = {
+            "27:9: "
+                + getCheckMessage(AbstractNameCheck.class,
+                    MSG_INVALID_PATTERN, "line_length", "^[a-z][a-zA-Z0-9]*$"),
+           };
+
+        verifySuppressedWithParser(getPath("InputSuppressionCommentFilterSuppressById.java"),
+                expectedViolation, suppressedViolation);
+    }
+    @Test
+    public void testCaching() throws Exception {
+        File file = new File(getPath("InputSuppressionCommentFilterSuppressById.java"));
+        DetailAST rootast = JavaParser.parseFile(file, JavaParser.Options.WITH_COMMENTS);
+        String[] lines = {"//CSOFF", "//CSON"};
+        final FileContents contents = new FileContents(
+                new FileText(file, Arrays.asList(lines)));
+        for (int lineNo = 0; lineNo < lines.length; lineNo++) {
+            final int colNo = lines[lineNo].indexOf("//");
+            if (colNo >= 0) {
+                contents.reportSingleLineComment(lineNo + 1, colNo);}
+        }
+        CommentListener exceptionalListener = new ExceptionalCommentListener();
```

Comment:
exceptional should be `final FileContents contents = new FileContents(` to throw exception on extra call of `contents.getSingleLineComments()` that Filter might do.

make ExceptionalFileContents class that throw excpeption only on second attempt of Filter to get comment.
`tagSuppressions();` in filter should be executed ones for two violations from the same file.

Reply:
@Kevin222004 , do you understand what I am suggesting ?

Reply:
no way to extend as such class is final:
`public final class FileContents implements CommentListener {`